imsApp/src/imsRecord.cc

```diff
@@ -3440,42 +3438,102 @@ static void flush_asyn( struct ims_info *mInfo )
-    size_t       nwrite;
-
-    sprintf( local_buf, "%s\r", msg );
+    int r = mInfo->reqH;
+    struct ims_asyn_req *req;
+
+    if (r == -1) /* This should never happen.  Famous last... */
+        return NULL;
+    req = &mInfo->req[r];
+    mInfo->reqH = req->link;
+    return req;
+}
 
-    if ( sEvt == 1 ) mInfo->sEvent->wait( 1.5 );
+/* Assume rMutex is locked! */
+static void free_request_buf ( struct ims_info *mInfo, struct ims_asyn_req *req )
+{
+    int r = req - &mInfo->req[0];
+    req->state = IAR_EMPTY;
+    req->link = mInfo->reqH;
+    mInfo->reqH = r;
+}
 
-    pasynOctetSyncIO->write( mInfo->pasynUser, local_buf, strlen(local_buf),
-                             timeout, &nwrite );
+static long send_msg( struct ims_info *mInfo, char const *msg )
+{
+    struct ims_asyn_req *req;
 
-    if ( sEvt == 1 ) mInfo->sEvent->signal();
+    mInfo->rMutex->lock();
+    req = get_request_buf(mInfo);
+    if (!req) {
+	mInfo->rMutex->unlock();
+	return -1;
+    }
 
+    sprintf(req->out, "%s\r\n", msg);
+    req->outsize = strlen(req->out);
+    req->state = IAR_SENDONLY;
+    mInfo->rMutex->unlock();
     return( 0 );
 }
 
 /******************************************************************************/
-static long recv_reply( struct ims_info *mInfo, char *rbbuf )
+/*
+ * The new regime: all of our PR statements in *code* start 'PR "XY=",' for some
+ * unique characters XY.  However... since we allow users to just send arbitrary
+ * commands, we can't really count on *all* of the commands looking like this.
+ * But if it *doesn't* look like this, we know it has to be a user command!
+ */
+static long send_recv_reply( struct ims_info *mInfo, char const *msg, char *rbbuf)
 {
-    const double timeout  = 1.0;
-    size_t       nread    = 0;
-    asynStatus   asyn_rtn = asynError;
-    int          eomReason;
-
-    asyn_rtn = pasynOctetSyncIO->read( mInfo->pasynUser, rbbuf, MAX_MSG_SIZE,
-                                       timeout, &nread, &eomReason );
+    struct ims_asyn_req *req;
+    int status;
+
+    mInfo->rMutex->lock();
+    req = get_request_buf(mInfo);
+    if (!req) {
+	mInfo->rMutex->unlock();
+	*rbbuf = 0;
+	return 0;
+    }
 
-    if ( (asyn_rtn != asynSuccess) || (nread <= 0) )
-    {
-        rbbuf[0] = '\0';
-        nread    = 0;
+    sprintf(req->out, "%s\r\n", msg);
+    req->outsize = strlen(req->out);
+    req->state = IAR_SEND;
+    /*
+     * We wouldn't be here if msg didn't start "PR ". Only question is
+     * if we are printing a literal string after this.
+     *
+     * We're trying to be safe here in case the user types something
+     * like 'PR ""' or 'PR "X"'.
+     */
+    if (msg[3] == '"') {
+	req->exp[0] = msg[4] == '"' ? 0 : msg[4];
+	req->exp[1] = msg[5] == '"' ? 0 : msg[5];
+	req->exp[2] = 0;
+    } else {
+	req->exp[0] = 0;
     }
+    req->rbbuf = rbbuf;
+    req->state = IAR_SEND;
+    mInfo->rMutex->unlock();
+
+    status = req->rbEvent->wait(1.0);
+    mInfo->rMutex->lock();
```

Comment:
Why are we allowed to unlock here before waiting for the response? Is there no way for us to receive the wrong message if we send two messages at once from different threads? Or are we simply always processing these in such a way that the order doesn't matter?

Reply:
I am using rMutex to control access to the request array.  If I don't unlock, io_controller will never be able to process anything!

Reply:
Ahh I see some of my misunderstanding now

Reply:
This goes back to @joshc-slac's comment about my paranoia, above.  The pthread library has a lock parameter for their semaphore wait... the *library* unlocks while waiting, and relocks when returning.

Sigh.